adonis-typings/drive.ts

```diff
@@ -47,10 +50,99 @@ declare module '@ioc:Adonis/Core/Drive' {
+     */
+    location: string
+
+    /**
+     * Flag to know if item represents file or directory
+     */
+    isFile: boolean
+
+    /**
+     * Original list item returned from underlying driver
+     */
+    original: T
+  }
+
+  /**
+   * Shape of the directory listing async iterable returned from list allowing to transform listing.
+   * This can be iterated directly using for-await-of loop or it can be converted to array using toArray().
+   */
+  export interface DirectoryListingContract<Driver extends DriverContract, T>
+    extends AsyncIterable<T> {
+    /**
+     * Reference to the driver for which the listing was created.
+     */
+    driver: Driver
+
+    /**
+     * Filter generated items of listing with the given predicate function.
+     */
+    filter(
+      predicate: (item: T, index: number, driver: Driver) => Promise<boolean> | boolean
+    ): DirectoryListingContract<Driver, T>
+
+    /**
+     * Transform generated items of listing with the given mapper function.
+     */
+    map<M>(
+      mapper: (item: T, index: number, driver: Driver) => Promise<M> | M
+    ): DirectoryListingContract<Driver, M>
+
+    /**
+     * Do recursive listing of items. Without the next function it will do listing of leaf nodes only.
+     * For advanced usage you can pass the next function which will get as parameter current item and it should
+     * return the next location for list or null if the recursion should stop and yield the current item.
+     * For advanced usage you can also limit the depth of recursion using the second argument of next function.
+     */
+    recursive(
+      next?: (current: T, depth: number, driver: Driver) => Promise<string | null> | string | null
+    ): DirectoryListingContract<Driver, T>
+
+    /**
+     * Add a piping chain function which gets the current async iterable and returns
+     * new async iterable with modified directory listing output.
+     * Function this is bound to instance of driver for which the listing is generated.
+     * This allows using async generator functions and reference the driver methods easily.
+     * Piping will always return clone of the current instance and add the function
+     * to the chain of new cloned instance only to prevent side effects.
+     */
+    pipe<U>(
+      fn: (this: Driver, source: AsyncIterable<T>) => AsyncIterable<U>
+    ): DirectoryListingContract<Driver, U>
+
+    /**
+     * Get the final async iterable after passing directory listing through chain of piping functions modifying the output.
+     */
+    toIterable(): AsyncIterable<T>
+
+    /**
+     * Convert directory listing to array.
+     */
+    toArray(): Promise<T[]>
+  }
+
+  /**
+   * Shape of the directory listing constructor, we export the constructor for others to add macros and getters to the class.
+   */
+  export interface DirectoryListingConstructorContract
+    extends MacroableConstructorContract<DirectoryListingContract<DriverContract, DriveListItem>> {
```

Comment:
Do we want directory listing class to be macroable?

Reply:
I think it is a good idea for good DX to allow chainable one line calls which are readable. For example, maybe somebody wants to list files matching a glob pattern, so he can add a nice chainable method and use it like this:

```typescript
// get json files in directory
const files = await driver.list('some/dir').glob('*.json').toArray()
```